### DISCUSSION TITLE
Multiple project environnements

### DISCUSSION THREAD

**Discussion by wlarch** [envault/envault#23] [Ideas]
Greetings, thank you for developing this wonderful tool for the Laravel community. We have previously developed a similar tool using AWS S3 Buckets and command line push/pull commands. Our tool allow pushing `.env` files to project S3 Buckets depending on the environment (local, staging, production) and  pulling `.env` files for different environments during deployment pipelines or manually using our local tool.

We understand your tool help managing local `.env` files.

Would you recommend using it for different environments (staging, production) ?

**If yes**, we would  create an Envault project for every project environment (my-app-staging, my-app-prod) and add the `.envault.json` file to `.gitignore` ?

**If not**, is it something you gave a thought ?

**Comment by danharrin** [comment]
Hey @wlarch! Yes, we've considered multiple-environment support before. Would love to hear your thoughts...
- What is your use case for syncing production and staging environments, since there is traditionally only one configuration source of each of these environments?
- Are you comfortable to store production credentials in a database? They are encrypted, but can be unencrypted easily if an attacker gained access to your `APP_KEY`.
- It would probably be preferable to allow apps to store multiple environments, instead of creating an app-per-environment. This works right now, but might not be the cleanest solution.

**Reply by wlarch** [reply]
Hey @danharrin, thanks for your prompt reply.
* We consider modifying `.env` files directly in Vapor not optimal for our staging and production environments. Instead, we use `.env` file syncing on deployment using Bitbucket Pipelines. At the moment our deployment process is executed as follows :
  1. `aws s3 cp $S3_BUCKET/.env.prod ./.env`
  2. `vapor env:push prod`
  3. `vapor deploy production`
* Thus always using the last version of the `.env` file present for the environment in our centralized source of truth (S3 Bucket).
* Security is the main concern, I will have to brainstorm with my colleagues about it. I suppose there could be an additional security layer (authentication/encryption for example).
* Yes, multiple environments for one Envault project would be best. Using the following syntax for example : 
  `npx envault domain.com 1 dev 5axa6haxhh8a `

**Reply by danharrin** [reply]
Perfect, let me know what you conclude on the security side of this. If you can present a good argument as to why this is secure / propose extra security features, I’m happy to look at developing this feature so the use case is officially supported. 👍

**Comment by ralphjsmit** [comment]
I also came across this and it's something that is in my opinion essential for how useful Envault is. In it's current state, Envault is a tool to manage and sync between multiple local environments. That's of course fine, but if Envaults supports multiple environments, it would turn into something way more useful: the nr. 1 tool to keep track of all your environment variables, in all your environments.

I didn't dive too deep into the source code, but I did look at the migrations and I think it can be relatively simple to add support for it.

Create a database (+ model) for managing Environments and relate it to `App` (like this: `$table->foreignId('app_id');`).
Add a simple admin page in Envault where users can add / manage environments for a given app. For each environment, store the name + a(n automatically generated) slug. 

If you add a migration for the `variables` table:
```
$table->foreignId('environment_id')->nullable()->constrained()->onUpdate('cascade')->onDelete('cascade');
```
If the `environment_id` is null, it applies to every environment. And if an `environment_id` is set, it applies only on one environment.
(To make this more flexible and allow to set variables to more than 1 environment, we can use a pivot table with a many-to-many relationship. Perhaps this is even better, because it also makes the need for NULLABLE redundant. And the relationship later between App and Environment becomes cleaner and easier.)

Slightly modify the `App->variables()`  to support some sort of where() clause.

What do you think?
